README: mention RemoveCorruptedOutput with an example

anilbey opened the pull request:
Updated README to provide an example for the RemoveCorruptedOutputMixin.

adrien-berchet commented:
I think we could also improve the tests of this feature a little bit. For example we could use `pytest.raises` here: https://github.com/BlueBrain/luigi-tools/blob/main/tests/test_task.py#L851

anilbey commented on README.md:
```diff
@@ -115,6 +117,32 @@ luigi -m my_module mytask --rerun
+
+When a task fails unexpectedly, it may leave an incomplete or corrupted output 
+that leads to wrong results in the downstream. With the RemoveCorruptedOutputMixin,
+Luigi automatically removes the output of the tasks that failed. This is the default behaviour 
```
good point

anilbey commented on README.md:
```diff
@@ -115,6 +117,32 @@ luigi -m my_module mytask --rerun
+class ResilientTask(RemoveCorruptedOutputMixin, luigi.Task):
+    """Resilient to the bugs caused by incomplete and corrupted outputs."""
+
+    RemoveCorruptedOutputMixin.clean_failed = luigi.BoolParameter(
+        significant=False,
+        default=True,
+        description="Trigger to remove the outputs of the failed tasks.",
+        parsing=luigi.BoolParameter.EXPLICIT_PARSING,
+    )
```
The default behaviour was True in my first PR. Though you convinced me to make it optional. I don't remember what the motivation was.

anilbey replied:
> Maybe it would be simpler to remove this and show an example of CLI command that triggers this behavior? 

Then there is also no need to have the ResilientTask class, right? 

TaskA can directly use the mixin like this.

`class TaskA(RemoveCorruptedOutputMixin, luigi.Task):`

adrien-berchet replied:
Yes exactly.

adrien-berchet replied:
> The default behaviour was True in my first PR. Though you convinced me to make it optional. I don't remember what the motivation was.

Ahah ok, I don't remember what was the argument. I am going to find out.

adrien-berchet replied:
Ah it's an old PR that is no more accessible on Gerrit :'(
I think the point was that the default behavior should stick to the default Luigi behavior to not confuse the user. And as you showed in your example, it is still easy to reverse the behavior if someone wants to.
So I think that having a simpler example with the luigi command is enough. Do you agree?

anilbey replied:
Ok sounds good.

adrien-berchet replied:
Ah yes. And such default also makes things easier for a user to send the failed state to the support.

anilbey commented:
> I think we could also improve the tests of this feature a little bit. For example we could use `pytest.raises` here: https://github.com/BlueBrain/luigi-tools/blob/main/tests/test_task.py#L851

Hmm, something fishy is happening with the tests. The tests work as excepted but the Luigi exception is no longer caught.

anilbey commented:
Ok, the exception gets handled by Luigi. I removed the try blocks and added extra assert statements that check the `complete()` method of the task instance.

adrien-berchet commented on README.md:
```diff
@@ -127,20 +127,17 @@ in other workflow management systems such as [Snakemake][snakemake_url].
-    """TaskA removes its output upon failure."""
-    pass
+The `clean_failed` flag must explicitly be set to true.
+This allows users to set it to false to debug the output without changing the code.
```
I think it's worth clarifying  that the default value is `False`.

anilbey replied:
Done

adrien-berchet commented:
> Ok, the exception gets handled by Luigi. I removed the try blocks and added extra assert statements that check the `complete()` method of the task instance.

Ah yes ok. If you want to see an example on how to check the exception raised inside a task you can look here: https://github.com/BlueBrain/luigi-tools/blob/main/tests/test_task.py#L420

But it's not the main point of this feature so it's as you prefer.

anilbey commented:
> > Ok, the exception gets handled by Luigi. I removed the try blocks and added extra assert statements that check the `complete()` method of the task instance.
> 
> Ah yes ok. If you want to see an example on how to check the exception raised inside a task you can look here: https://github.com/BlueBrain/luigi-tools/blob/main/tests/test_task.py#L420
> 
> But it's not the main point of this feature so it's as you prefer.

Thanks!
 I added it nonetheless to make sure no other exception but our intended exception that is thrown and caught.

adrien-berchet reviewed (approved):
Looks good to me, thanks!
